File: table/src/main/java/com/adsamcik/table/Table.java
package com.adsamcik.table;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.support.v4.graphics.ColorUtils;
import android.support.v7.widget.CardView;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.Pair;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Locale;

import static com.adsamcik.table.Util.dpToPx;
import static com.adsamcik.table.Util.formatNumber;
import static com.adsamcik.table.Util.getAccentColor;
import static com.adsamcik.table.Util.getDisplayMetrics;
import static com.adsamcik.table.Util.getPressedColorRippleDrawable;

public class Table {
	public @AppendBehavior
	int appendBehavior;

	public String getTitle() {
		return title;
	}

	private String title = null;
	private ArrayList<Pair<String, String>> data;
	private ArrayList<Pair<String, View.OnClickListener>> buttons = null;

	private boolean showNumber;

	private int wrapperMarginDp;

	private int dividerColor = 0;

	/**
	 * Table constructor
	 *
	 * @param rowCount   number of data (used to initialize array holding data)
	 * @param showNumber show number of row (starts at 1)
	 */
	public Table(@NonNull Context context, int rowCount, boolean showNumber, int wrapperMarginDp, @AppendBehavior int appendBehavior) {
		this.data = new ArrayList<>(rowCount);
		this.showNumber = showNumber;
		this.appendBehavior = appendBehavior;
		this.wrapperMarginDp = wrapperMarginDp;
	}

	private void updateDividerColor(@NonNull CardView cardView) {
		int color = cardView.getCardBackgroundColor().getDefaultColor();
		double lum = ColorUtils.calculateLuminance(color);

		if (lum > 0.5)
			dividerColor = Color.argb(30, 0, 0, 0);
		else
			dividerColor = Color.argb(30, 255, 255, 255);
	}

	/*public void addToViewGroup(@NonNull ViewGroup viewGroup, @NonNull Context context, int index, boolean animate, long delay) {
		if (index >= 0 && index < viewGroup.getChildCount())
			viewGroup.addView(view, index);
		else
			viewGroup.addView(view);

		if (animate) {
			view.setTranslationY(viewGroup.getHeight());
			view.setAlpha(0);
			view.animate()
					.translationY(0)
					.setInterpolator(new DecelerateInterpolator(3.f))
					.setDuration(700)
					.setStartDelay(delay)
					.alpha(1)
					.start();
		}
	}*/

	/**
	 * Sets single title for whole table
	 *
	 * @param title title
	 * @return this table
	 */
	public Table setTitle(String title) {
		this.title = title;
		return this;
	}

	/**
	 * Add button to the bottom of the table
	 *
	 * @param text     title of the button
	 * @param callback on click callback
	 * @return this table
	 */
	public Table addButton(String text, View.OnClickListener callback) {
		if (buttons == null)
			buttons = new ArrayList<>(2);
		buttons.add(new Pair<>(text, callback));
		return this;
	}

	/**
	 * Adds data to 2 columns on the last row, only use this with 2 columns (+1 if row numbering is enabled)
	 *
	 * @param name  row name
	 * @param value row value
	 * @return this table
	 */
	public Table addData(String name, String value) {
		data.add(new Pair<>(name, value));
		return this;
	}

	private TableRow generateButtonsRow(@NonNull Context context, @StyleRes int theme, int sideMargin) {
		if (buttons != null) {
			DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
			TableRow row = new TableRow(context);
			TableLayout.LayoutParams lp = new TableLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
			lp.topMargin = dpToPx(displayMetrics, 4);
			lp.setMargins(sideMargin, 0, sideMargin, 0);
			row.setLayoutParams(lp);

			for (int i = 0; i < buttons.size(); i++)
				row.addView(generateButton(context, displayMetrics, i, theme));
			return row;
		}
		return null;
	}

	private TextView generateButton(@NonNull Context context, DisplayMetrics displayMetrics, int index, @StyleRes int theme) {
		Button button = new Button(context, null, theme);

		//value caching
		int dp48px = dpToPx(displayMetrics, 48);
		int dp16px = dpToPx(displayMetrics, 16);

		//dimensions
		button.setMinWidth(dp48px);
		button.setPadding(dp16px, 0, dp16px, 0);
		button.setHeight(dp48px);

		//layout params
		TableRow.LayoutParams layoutParams = new TableRow.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
		layoutParams.setMargins(0, dp16px, 0, 0);
		button.setLayoutParams(layoutParams);

		//style
		button.setText(buttons.get(index).first.toUpperCase());
		button.setTypeface(Typeface.defaultFromStyle(Typeface.BOLD));
		button.setOnClickListener(buttons.get(index).second);
		button.setTextSize(16);
		button.setGravity(Gravity.CENTER);
		button.setBackground(getPressedColorRippleDrawable(0, getAccentColor(context), context.getDrawable(R.drawable.rectangle)));


		return button;
	}

	private TableRow generateDataRow(@NonNull Context context, int index, @StyleRes int theme) {
		TableRow row = new TableRow(context);

		if (showNumber) {
			TextView rowNum = new TextView(context, null, theme);
			rowNum.setText(String.format(Locale.UK, "%d", index + 1));
			rowNum.setLayoutParams(new TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 0.5f));
			row.addView(rowNum);
		}

		TextView textId = new TextView(context, null, theme);
		textId.setText(data.get(index).first);
		textId.setLayoutParams(new TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 3f));
		row.addView(textId);

		TextView textValue = new TextView(context, null, theme);
		String value = data.get(index).second;
		try {
			textValue.setText(formatNumber(Integer.parseInt(value)));
		} catch (NumberFormatException e) {
			textValue.setText(value);
		}
		textValue.setLayoutParams(new TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 2f));
		textValue.setGravity(Gravity.END);
		row.addView(textValue);

		return row;
	}

	/**
	 * Generates view for the table
	 *
	 * @param context        context
	 * @param requireWrapper FrameView wrapper for margin
	 * @return card view with the new table
	 */
	public View getView(@NonNull Context context, @Nullable View recycle, boolean requireWrapper, @StyleRes int theme) {
		Resources r = context.getResources();
		boolean addWrapper = wrapperMarginDp != 0 || requireWrapper;

		CardView cardView = null;
		FrameLayout frameLayout = null;

		DisplayMetrics displayMetrics = getDisplayMetrics(context);

		if (recycle != null) {
			if (recycle instanceof CardView) {
				cardView = (CardView) recycle;
			} else if (recycle instanceof FrameLayout) {
				View viewTest = ((FrameLayout) recycle).getChildAt(0);
				if (viewTest instanceof CardView) {
					cardView = (CardView) viewTest;
					frameLayout = (FrameLayout) recycle;
					//For some reason there can be view or views that have HW acceleration disabled
					//and without it shadows are not drawn. This solution recreates those views and fixes the issue
					if (!frameLayout.isHardwareAccelerated()) {
						((FrameLayout) recycle).removeView(cardView);
						addWrapper = true;
					} else
						addWrapper = false;
				}
			}
		}

		if (cardView == null) {
			cardView = new CardView(context, null, theme);
			updateDividerColor(cardView);
		}

		final int padding = (int) r.getDimension(R.dimen.table_padding);
		final int itemVerticalPadding = (int) r.getDimension(R.dimen.table_item_vertical_padding);

		TableLayout layout = (TableLayout) cardView.getChildAt(0);
		if (layout == null) {
			layout = new TableLayout(context);
			layout.setPadding(0, 30, 0, 30);
			cardView.addView(layout);
		} else
			layout.removeAllViews();

		if (title != null) {
			TextView titleView = (TextView) layout.getChildAt(0);
			if (titleView == null) {
				titleView = new TextView(context, null, theme);
				titleView.setTextSize(18);
				titleView.setTypeface(null, Typeface.BOLD);
				titleView.setGravity(Gravity.CENTER);
				titleView.setLines(1);
				titleView.setEllipsize(TextUtils.TruncateAt.END);

				TableLayout.LayoutParams titleLayoutParams = new TableLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
				titleLayoutParams.setMargins(padding, 0, padding, dpToPx(displayMetrics, 16));
				titleView.setLayoutParams(titleLayoutParams);

				layout.addView(titleView, 0);
			}

			titleView.setText(title);
		}

		if (data.size() > 0) {
			TableRow row = generateDataRow(context, 0, theme);
			row.setPadding(padding, itemVerticalPadding, padding, itemVerticalPadding);
			layout.addView(row);
			if (data.size() > 1) {
				TableLayout.LayoutParams layoutParams = new TableLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dpToPx(displayMetrics, 1));

				for (int i = 1; i < data.size(); i++) {
					View divider = new View(context);
					divider.setLayoutParams(layoutParams);
					divider.setBackgroundColor(dividerColor);
					layout.addView(divider);

					row = generateDataRow(context, i, theme);
					row.setPadding(padding, itemVerticalPadding, padding, itemVerticalPadding);
					layout.addView(row);
				}
			}
		}

		TableRow buttonsRow = generateButtonsRow(context, theme, padding);
		if (buttonsRow != null)
			layout.addView(buttonsRow);

		if (addWrapper) {
			frameLayout = new FrameLayout(context);

			if (wrapperMarginDp != 0) {
				FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
				int margin = dpToPx(displayMetrics, this.wrapperMarginDp);
				layoutParams.setMargins(margin, margin, margin, margin);
				cardView.setLayoutParams(layoutParams);
			}
			frameLayout.addView(cardView);
		}

		return frameLayout != null ? frameLayout : cardView;
	}
}

File: table/src/main/java/com/adsamcik/table/Table.kt
package com.adsamcik.table

import android.content.Context
import android.content.res.Resources
import android.graphics.Color
import android.graphics.Typeface
import android.os.Build
import android.support.annotation.StyleRes
import android.support.v4.graphics.ColorUtils
import android.support.v7.widget.CardView
import android.text.TextUtils
import android.util.DisplayMetrics
import android.util.Log
import android.util.Pair
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView

import java.util.ArrayList
import java.util.Locale

import com.adsamcik.table.Util.dpToPx
import com.adsamcik.table.Util.formatNumber
import com.adsamcik.table.Util.getAccentColor
import com.adsamcik.table.Util.getDisplayMetrics
import com.adsamcik.table.Util.getPressedColorRippleDrawable

class Table
/**
 * Table constructor
 *
 * @param rowCount   number of data (used to initialize array holding data)
 * @param showNumber show number of row (starts at 1)
 */
(rowCount: Int, private val showNumber: Boolean, private val wrapperMarginDp: Int, @param:AppendBehavior @field:AppendBehavior
var appendBehavior: Int) {

    private var title: String? = null
    private val data: ArrayList<Pair<String, String>>
    private var buttons: ArrayList<Pair<String, View.OnClickListener>>? = null

    private var dividerColor = 0

    fun getTitle(): String? {
        return title
    }

    init {
        this.data = ArrayList(rowCount)
    }

    private fun updateDividerColor(cardView: CardView) {
        val color = cardView.cardBackgroundColor.defaultColor
        val lum = ColorUtils.calculateLuminance(color)

        if (lum > 0.5)
            dividerColor = Color.argb(30, 0, 0, 0)
        else
            dividerColor = Color.argb(30, 255, 255, 255)
    }

    /*public void addToViewGroup(@NonNull ViewGroup viewGroup, @NonNull Context context, int index, boolean animate, long delay) {
		if (index >= 0 && index < viewGroup.getChildCount())
			viewGroup.addView(view, index);
		else
			viewGroup.addView(view);

		if (animate) {
			view.setTranslationY(viewGroup.getHeight());
			view.setAlpha(0);
			view.animate()
					.translationY(0)
					.setInterpolator(new DecelerateInterpolator(3.f))
					.setDuration(700)
					.setStartDelay(delay)
					.alpha(1)
					.start();
		}
	}*/

    /**
     * Sets single title for whole table
     *
     * @param title title
     * @return this table
     */
    fun setTitle(title: String): Table {
        this.title = title
        return this
    }

    /**
     * Add button to the bottom of the table
     *
     * @param text     title of the button
     * @param callback on click callback
     * @return this table
     */
    fun addButton(text: String, callback: View.OnClickListener): Table {
        if (buttons == null)
            buttons = ArrayList<Pair<String, OnClickListener>>(2)
        buttons!!.add(Pair<String, OnClickListener>(text, callback))
        return this
    }

    /**
     * Adds data to 2 columns on the last row, only use this with 2 columns (+1 if row numbering is enabled)
     *
     * @param name  row name
     * @param value row value
     * @return this table
     */
    fun addData(name: String, value: String): Table {
        data.add(Pair(name, value))
        return this
    }

    private fun generateButtonsRow(context: Context, @StyleRes theme: Int, sideMargin: Int): TableRow? {
        if (buttons != null) {
            val displayMetrics = context.resources.displayMetrics
            val row = TableRow(context)
            val lp = TableLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            lp.topMargin = dpToPx(displayMetrics, 4)
            lp.setMargins(sideMargin, 0, sideMargin, 0)
            row.layoutParams = lp

            for (i in buttons!!.indices)
                row.addView(generateButton(context, displayMetrics, i, theme))
            return row
        }
        return null
    }

    private fun generateButton(context: Context, displayMetrics: DisplayMetrics, index: Int, @StyleRes theme: Int): TextView {
        val button = Button(context, null, theme)

        //value caching
        val dp48px = dpToPx(displayMetrics, 48)
        val dp16px = dpToPx(displayMetrics, 16)

        //dimensions
        button.minWidth = dp48px
        button.setPadding(dp16px, 0, dp16px, 0)
        button.height = dp48px

        //layout params
        val layoutParams = TableRow.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        layoutParams.setMargins(0, dp16px, 0, 0)
        button.layoutParams = layoutParams

        //style
        button.setText(buttons!![index].first.toUpperCase())
        button.typeface = Typeface.defaultFromStyle(Typeface.BOLD)
        button.setOnClickListener(buttons!![index].second)
        button.textSize = 16f
        button.gravity = Gravity.CENTER
        button.background = getPressedColorRippleDrawable(0, getAccentColor(context), context.getDrawable(R.drawable.rectangle))


        return button
    }

    private fun generateDataRow(context: Context, index: Int, @StyleRes theme: Int): TableRow {
        val row = TableRow(context)

        if (showNumber) {
            val rowNum = TextView(context, null, theme)
            rowNum.setText(String.format(Locale.UK, "%d", index + 1))
            rowNum.layoutParams = TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 0.5f)
            row.addView(rowNum)
        }

        val textId = TextView(context, null, theme)
        textId.text = data[index].first
        textId.layoutParams = TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 3f)
        row.addView(textId)

        val textValue = TextView(context, null, theme)
        val value = data[index].second
        try {
            textValue.text = formatNumber(Integer.parseInt(value))
        } catch (e: NumberFormatException) {
            textValue.text = value
        }

        textValue.layoutParams = TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 2f)
        textValue.gravity = Gravity.END
        row.addView(textValue)

        return row
    }

    /**
     * Generates view for the table
     *
     * @param context        context
     * @param requireWrapper FrameView wrapper for margin
     * @return card view with the new table
     */
    fun getView(context: Context, recycle: View?, requireWrapper: Boolean, @StyleRes theme: Int): View {
        val r = context.resources
        var addWrapper = wrapperMarginDp != 0 || requireWrapper

        var cardView: CardView? = null
        var frameLayout: FrameLayout? = null

        val displayMetrics = getDisplayMetrics(context)

        if (recycle != null) {
            if (recycle is CardView) {
                cardView = recycle
            } else if (recycle is FrameLayout) {
                val viewTest = recycle.getChildAt(0)
                if (viewTest is CardView) {
                    cardView = viewTest
                    frameLayout = recycle
                    //For some reason there can be view or views that have HW acceleration disabled
                    //and without it shadows are not drawn. This solution recreates those views and fixes the issue
                    if (!frameLayout.isHardwareAccelerated) {
                        recycle.removeView(cardView)
                        addWrapper = true
                    } else
                        addWrapper = false
                }
            }
        }

        if (cardView == null)
            cardView = CardView(context, null, theme)

        if (dividerColor == 0)
            updateDividerColor(cardView)

        val padding = r.getDimension(R.dimen.table_padding).toInt()
        val itemVerticalPadding = r.getDimension(R.dimen.table_item_vertical_padding).toInt()

        var layout: TableLayout? = cardView.getChildAt(0) as TableLayout
        if (layout == null) {
            layout = TableLayout(context)
            layout.setPadding(0, 30, 0, 30)
            cardView.addView(layout)
        } else
            layout.removeAllViews()

        if (title != null) {
            var titleView: TextView? = layout.getChildAt(0) as TextView
            if (titleView == null) {
                titleView = TextView(context, null, theme)
                titleView.textSize = 18f
                titleView.setTypeface(null, Typeface.BOLD)
                titleView.gravity = Gravity.CENTER
                titleView.setLines(1)
                titleView.ellipsize = TextUtils.TruncateAt.END

                val titleLayoutParams = TableLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
                titleLayoutParams.setMargins(padding, 0, padding, dpToPx(displayMetrics, 16))
                titleView.layoutParams = titleLayoutParams

                layout.addView(titleView, 0)
            }

            titleView.text = title
        }

        if (data.size > 0) {
            var row = generateDataRow(context, 0, theme)
            row.setPadding(padding, itemVerticalPadding, padding, itemVerticalPadding)
            layout.addView(row)
            if (data.size > 1) {
                val layoutParams = TableLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dpToPx(displayMetrics, 1))

                for (i in 1 until data.size) {
                    val divider = View(context)
                    divider.layoutParams = layoutParams
                    divider.setBackgroundColor(dividerColor)
                    layout.addView(divider)

                    row = generateDataRow(context, i, theme)
                    row.setPadding(padding, itemVerticalPadding, padding, itemVerticalPadding)
                    layout.addView(row)
                }
            }
        }

        val buttonsRow = generateButtonsRow(context, theme, padding)
        if (buttonsRow != null)
            layout.addView(buttonsRow)

        if (addWrapper) {
            frameLayout = FrameLayout(context)

            if (wrapperMarginDp != 0) {
                val layoutParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
                val margin = dpToPx(displayMetrics, this.wrapperMarginDp)
                layoutParams.setMargins(margin, margin, margin, margin)
                cardView.layoutParams = layoutParams
            }
            frameLayout.addView(cardView)
        }

        return if (frameLayout != null) frameLayout else cardView
    }
}

